detox/src/devices/Device.js

```diff
@@ -5,119 +5,95 @@ const { traceCall } = require('../utils/trace');
 
 class Device {
   constructor({
+    appsConfig,
     behaviorConfig,
     deviceConfig,
     deviceDriver,
     emitter,
-    sessionConfig
+    sessionConfig,
+    runtimeErrorBuilder,
   }) {
+    this._appsConfig = appsConfig;
     this._behaviorConfig = behaviorConfig;
     this._deviceConfig = deviceConfig;
     this._sessionConfig = sessionConfig;
     this._emitter = emitter;
+    this._errorBuilder = runtimeErrorBuilder;
+
+    this._currentApp = null;
+    this._cachedAppLaunchArgs = null;
+    this._deviceId = undefined;
     this._processes = {};
-    this._launchArgs = new LaunchArgs(deviceConfig.launchArgs);
+
     this.deviceDriver = deviceDriver;
     this.deviceDriver.validateDeviceConfig(deviceConfig);
     this.debug = debug;
   }
 
-  async prepare() {
-    await this.deviceDriver.prepare();
-
-    this._deviceId = await traceCall('acquireDevice', () =>
-      this.deviceDriver.acquireFreeDevice(this._deviceConfig.device || this._deviceConfig.name));
-    this._bundleId = await this.deviceDriver.getBundleIdFromBinary(this._deviceConfig.binaryPath);
+  get id() {
+    return this._deviceId;
   }
 
-  async launchApp(params = {}, bundleId = this._bundleId) {
-    return traceCall('launchApp', () => this._doLaunchApp(params, bundleId));
+  get name() {
+    return this.deviceDriver.name;
   }
 
-  async _doLaunchApp(params, bundleId) {
-    const deviceId = this._deviceId;
-    const payloadParams = ['url', 'userNotification', 'userActivity'];
-    const hasPayload = this._assertHasSingleParam(payloadParams, params);
-    const newInstance = params.newInstance !== undefined
-      ? params.newInstance
-      : this._processes[bundleId] == null;
+  get type() {
+    return this._deviceConfig.type;
+  }
 
-    if (params.delete) {
-      await this._terminateApp();
-      await this._reinstallApp();
-    } else if (newInstance) {
-      await this._terminateApp();
+  get appLaunchArgs() {
+    if (!this._cachedAppLaunchArgs) {
```

Comment:
(as discussed 1x1) Consider revisiting the implicitness of this approach.
Maybe as a tradeoff, a mere rename (e.g `_currentAppArgs`) might help make it more understandable.
Or to solve altogether - what if a LaunchArgs object is created per each app, to replace the plain config? Just a suggestion.

Reply:
I'll add a TODO after the merge